pkg/runtime/function/cast_date.go

```diff
@@ -0,0 +1,232 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package function
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+import (
+	"github.com/pkg/errors"
+)
+
+import (
+	"github.com/arana-db/arana/pkg/proto"
+)
+
+// FuncCastDate is  https://dev.mysql.com/doc/refman/5.6/en/cast-functions.html#function_cast
+const FuncCastDate = "CAST_DATE"
+
+var DateSep = "~!@#$%^&*_+=:;,.|/?\\(\\)\\[\\]\\{\\}\\-\\\\"
+var _ proto.Func = (*castDateFunc)(nil)
+
+func init() {
+	proto.RegisterFunc(FuncCastDate, castDateFunc{})
+}
+
+type castDateFunc struct{}
+
+func (a castDateFunc) Apply(ctx context.Context, inputs ...proto.Valuer) (proto.Value, error) {
+	// expr
+	val, err := inputs[0].Value(ctx)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	dateArgs := fmt.Sprint(val)
+	if strings.Compare(dateArgs, "") == 0 {
+		return a.DefaultDateValue(), nil
+	}
+
+	// format - YY-MM-DD, YYYY-MM-DD
+	pat := "^\\d{1,4}[" + DateSep + "]+\\d{1,2}[" + DateSep + "]+\\d{1,2}$"
+	match, err := regexp.MatchString(pat, dateArgs)
+	if match && err == nil {
+		rep := regexp.MustCompile(`[` + DateSep + `]+`)
+		dateArgsReplace := rep.ReplaceAllStringFunc(dateArgs, func(s string) string { return "-" })
+		dateYear, dateMonth, dateDay := a.splitDateWithSep(dateArgsReplace)
+
+		if a.IsYearValid(dateYear) && a.IsMonthValid(dateMonth) && a.IsDayValid(dateYear, dateMonth, dateDay) {
+			dateStr := a.DateOutput(dateYear, dateMonth, dateDay)
+			return dateStr, nil
+		} else {
+			return a.DefaultDateValue(), nil
+		}
+	}
+	// format - YYYYMMDD, YYMMDD
+	pat = "^\\d{5,8}$"
+	match, err = regexp.MatchString(pat, dateArgs)
+	if match && err == nil {
+		dateYear, dateMonth, dateDay := a.splitDateWithoutSep(dateArgs)
+
+		if a.IsYearValid(dateYear) && a.IsMonthValid(dateMonth) && a.IsDayValid(dateYear, dateMonth, dateDay) {
+			dateStr := a.DateOutput(dateYear, dateMonth, dateDay)
+			return dateStr, nil
+		} else {
+			return a.DefaultDateValue(), nil
+		}
+	}
+
+	return a.DefaultDateValue(), nil
+}
+
+func (a castDateFunc) NumInput() int {
+	return 1
+}
+
+func (a castDateFunc) DateOutput(year, month, day int) proto.Value {
+	dayStr := fmt.Sprint(day)
+	if day < 10 {
+		dayStr = "0" + dayStr
+	}
+	monthStr := fmt.Sprint(month)
+	if month < 10 {
+		monthStr = "0" + monthStr
+	}
+	yearStr := fmt.Sprint(year)
+	if year >= 100 && year <= 999 {
+		yearStr = "0" + yearStr
+	}
+
+	dateStr := yearStr + "-" + monthStr + "-" + dayStr
+	return proto.NewValueString(dateStr)
+}
+
+func (a castDateFunc) splitDateWithSep(dateArgs string) (year, month, day int) {
+	dateLen := len(dateArgs)
+	dateDayStr := string(dateArgs[dateLen-1 : dateLen])
```

Comment:
I think there's no need to use `string(...)` again, please check it.

Reply:
done